src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -87,13 +128,97 @@ public override TypedExpression Transform(TypedExpression ex)
             return base.Transform(ex);
         }
 
-        private string CreateTypePrefix(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+        private Concretion CreateConcretion(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+        {
+            Concretion result = new Concretion();
+            foreach (var kvp in typeDict)
+            {
+                result.Add((kvp.Key.TypeName, kvp.Value.Resolution));
+            }
+            return result;
+        }
+
+        private Identifier GetConcreteIdentifier(Identifier.GlobalCallable globalCallable, ImmutableDictionary<QsTypeParameter, ResolvedType> types)
+        {
+            Concretion target = CreateConcretion(types);
+
+            // TODO: handle if globalCallable is not in Generics
+
+            var (original, concretions) = Generics[globalCallable.Item];
+
+            NonNullable<string> name;
+
+            if (concretions.ContainsKey(target))
+            {
+                name = concretions[target].FullName.Name;
+            }
+            else
+            {
+                // Create new name
+                string nameString = Guid.NewGuid().ToString() + "_" + globalCallable.Item.Name.Value;
+
+                // TODO: Check for identifier name collisions (not likely)
+                // - check in global namespace
+                // - check in concretion dict for current generic
+
+                name = NonNullable<string>.New(nameString);
+                
+                QsCallable newCallable = ReplaceTypeParamsSyntax.Apply(types, original)
+                    .WithFullName(fullname => new QsQualifiedName(globalCallable.Item.Namespace, name));
+
+                concretions.Add(target, newCallable);
+            }
+
+            return Identifier.NewGlobalCallable(new QsQualifiedName(globalCallable.Item.Namespace, name));
+        }
+    }
+
+    public class ResolveGenericsExpressionKind : ExpressionKindTransformation<ResolveGenericsExpression>
+    {
+        public ResolveGenericsExpressionKind(ResolveGenericsExpression expr) : base(expr) { }
+
+        public override QsExpressionKind<TypedExpression, Identifier, ResolvedType> onIdentifier(Identifier sym, QsNullable<ImmutableArray<ResolvedType>> tArgs)
+        {
+            return base.onIdentifier(sym, tArgs);
+        }
+    }
+
+    #endregion
+
+    #region ReplaceTypeParams
+
+    internal class ReplaceTypeParamsSyntax :
+        SyntaxTreeTransformation<MinorTransformations.ReplaceTypeParams>
+    {
+        public static QsCallable Apply(ImmutableDictionary<QsTypeParameter, ResolvedType> typeParams, QsCallable callable)
+        {
+            if (callable == null) throw new ArgumentNullException(nameof(callable));
+
+            var filter = new ReplaceTypeParamsSyntax(new MinorTransformations.ReplaceTypeParams(typeParams));
+
+            // Create a holding namespace to use transform on
```

Comment:
Why not directly call the transformation on the callable?

Reply:
That would be ideal. I haven't seen how to call a transformation on something other than a QsNamespace. Can you give me a little more feedback on how I would call a transformation on a callable? Being able to transform callables directly would simplify things quite a bit.

Reply:
Instead of filter.Transform you would just call the corresponding filter.onCallableImplementation. Maybe the naming here is not the best; it takes a QsCallable and returns a QsCallable.